internal/internal_task_handlers.go

```diff
@@ -1156,10 +1155,30 @@ func (w *workflowExecutionContextImpl) ResetIfStale(task *workflowservice.PollWo
 			tagCachedPreviousStartedEventID, w.previousStartedEventID,
 			tagTaskFirstEventID, task.History.Events[0].GetEventId(),
 			tagTaskStartedEventID, task.GetStartedEventId(),
-			tagPreviousStartedEventID, task.GetPreviousStartedEventId())
-
+			tagPreviousStartedEventID, task.GetPreviousStartedEventId(),
+		)
 		w.clearState()
-		return w.resetStateIfDestroyed(task, historyIterator)
+		err := w.resetStateIfDestroyed(task, historyIterator)
+		if err != nil {
+			return err
+		}
+		if w.workflowInfo != nil {
+			// Reset the search attributes and memos from the WorkflowExecutionStartedEvent.
+			// The search attributes and memo may have been modified by calls like UpsertMemo
+			// or UpsertSearchAttributes. They must be reset to avoid non determinism on replay.
+			h := task.History
+			startedEvent := h.Events[0]
+			attributes := startedEvent.GetWorkflowExecutionStartedEventAttributes()
+			// The first event in the history has to be WorkflowExecutionStartedEvent because
+			// w.clearState() should "destroy" the state and w.resetStateIfDestroyed() will reset the
+			// history of the execution context if it is considered "destroyed".
+			if attributes == nil {
+				return errors.New("first history event is not WorkflowExecutionStarted")
+			}
+			w.workflowInfo.SearchAttributes = attributes.SearchAttributes
+			w.workflowInfo.Memo = attributes.Memo
```

Comment:
I think this should be done in resetStateIfDestroyed() or even further in resetHistory().

Reply:
Done

Reply:
Can I get a bit more clarity as to "why" here? Specifically, what cases is it stale but not destroyed that we don't want to reset search attributes and memo?